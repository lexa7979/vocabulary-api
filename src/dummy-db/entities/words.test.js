const { getTestId, _testHelpers: TestIdHelpers } = require('../testIds');

const Words = require('./words');

const { ASYNC, bold, EXPECTS, IS_ACCESSIBLE, REJECTS, RESOLVES } = require('../../../test');

const Global = {};

describe('Test-data service "words"', () => {
  runTestsAboutGetWord();
  runTestsAboutGetWordWithTranslations();
  runTestsAboutListAllWords();
  runTestsAboutGetTranslation();
  runTestsAboutListAllTranslationsOfWord();
});

function runTestsAboutGetWord() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold('getWord()')} which`, () => {
    const { getWord } = Words;

    it(IS_ACCESSIBLE, () => expect(getWord).toBeFunction());

    it(`${EXPECTS} one argument (wordId)`, () => expect(getWord).toHaveLength(1));

    it(`- when used w/o arguments - ${REJECTS} as expected`, () =>
      expect(getWord).rejects.toThrow('invalid ID'));

    it(`- when used with valid word-ID - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const result = await getWord(getTestId('word1'));

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result);
      expect(copy).toMatchSnapshot(`> word1 <`);
    });

    it(`- when used with invalid word-ID - ${REJECTS} as expected`, () =>
      expect(() => getWord(getTestId('invalid'))).rejects.toThrow('invalid ID'));
  });
}

function runTestsAboutGetWordWithTranslations() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold(
    'getWordWithTranslations'
  )} which`, () => {
    const { getWordWithTranslations } = Words;

    it(IS_ACCESSIBLE, () => expect(getWordWithTranslations).toBeFunction());

    it(`${EXPECTS} one argument (wordId)`, () => expect(getWordWithTranslations).toHaveLength(1));

    it(`- when used w/o argument - ${REJECTS} as expected`, () =>
      expect(getWordWithTranslations).rejects.toThrow('invalid ID'));

    it(`- when used with valid word-ID - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const result = await getWordWithTranslations(getTestId('word2'));

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result);
      expect(copy).toMatchSnapshot(`> word2 + translations <`);
    });

    it(`- when used with invalid word-ID - ${REJECTS} as expected`, () =>
      expect(() => getWordWithTranslations(getTestId('invalid'))).rejects.toThrow('invalid ID'));
  });
}

function runTestsAboutListAllWords() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold('listAllWords()')} which`, () => {
    const { listAllWords } = Words;

    it(IS_ACCESSIBLE, () => expect(listAllWords).toBeFunction());

    it(`${EXPECTS} no argument`, () => expect(listAllWords).toHaveLength(0));

    it(`- when used w/o arguments - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const result = await listAllWords();

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result);
      expect(copy).toMatchSnapshot(`> all words <`);
    });

    it(`- when used with argument "first" - ${RESOLVES} as expected`, async () => {
      const result = await listAllWords(4, undefined);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result).map(item => item.id);
      expect(copy).toMatchInlineSnapshot(`
        Array [
          "(ID:word1)",
          "(ID:word2)",
          "(ID:word3)",
          "(ID:word4)",
        ]
      `);
    });

    it(`- when used with argument "offset" - ${RESOLVES} as expected`, async () => {
      const result = await listAllWords(undefined, 8);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result).map(item => item.id);
      expect(copy).toMatchInlineSnapshot(`
        Array [
          "(ID:word9)",
          "(ID:word10)",
          "(ID:word11)",
          "(ID:word12)",
        ]
      `);
    });
  });
}

function runTestsAboutGetTranslation() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold('getTranslation()')} which`, () => {
    const { getTranslation } = Words;

    it(IS_ACCESSIBLE, () => expect(getTranslation).toBeFunction());

    it(`${EXPECTS} one argument (translationId)`, () => expect(getTranslation).toHaveLength(1));

    it(`- when used w/o argument - ${REJECTS} as expected`, () =>
      expect(getTranslation).rejects.toThrow('invalid ID'));

    it(`- when used with a valid ID - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const id = getTestId('word1-translation1');

      const result = await getTranslation(id);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result);
      expect(copy).toMatchSnapshot(`> word1-translation1 <`);
    });

    it(`- when used with an invalid ID - ${REJECTS} as expected`, () =>
      expect(() => getTranslation(getTestId('invalid'))).rejects.toThrow('invalid ID'));
  });
}

function runTestsAboutListAllTranslationsOfWord() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold(
    'listAllTranslationsOfWord()'
  )} which`, () => {
    const { listAllTranslationsOfWord } = Words;

    it(IS_ACCESSIBLE, () => expect(listAllTranslationsOfWord).toBeFunction());

    it(`${EXPECTS} one argument (wordId)`, () => expect(listAllTranslationsOfWord).toHaveLength(1));

    it(`- when used w/o argument - ${RESOLVES} with empty array`, () =>
      expect(listAllTranslationsOfWord()).resolves.toEqual([]));

    it(`- when used with a valid word-ID - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const results = await listAllTranslationsOfWord(getTestId('word1'));

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(results).map(item => [
        item.id,
        item.text_de,
        item.text_sv,
      ]);
      expect(copy).toMatchSnapshot(`> list of translations <`);
    });

    it(`- when used with valid ID and argument "first" - ${RESOLVES} as expected`, async () => {
      const result = await listAllTranslationsOfWord(getTestId('word1'), 3, undefined);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result).map(item => item.id);
      expect(copy).toMatchInlineSnapshot(`
        Array [
          "(ID:word1-translation1)",
          "(ID:word1-translation2)",
          "(ID:word1-translation3)",
        ]
      `);
    });

    it(`- when used with valid ID and argument "offset" - ${RESOLVES} as expected`, async () => {
      const result = await listAllTranslationsOfWord(getTestId('word1'), undefined, 3);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result).map(item => item.id);
      expect(copy).toMatchInlineSnapshot(`
        Array [
          "(ID:word1-translation4)",
          "(ID:word1-translation5)",
          "(ID:word1-translation6)",
          "(ID:word1-translation7)",
        ]
      `);
    });

    it(`- when used with an invalid word-ID - ${RESOLVES} with empty array`, () =>
      expect(listAllTranslationsOfWord(getTestId('invalid'))).resolves.toEqual([]));
  });
}

function _nextListHint() {
  Global.nextListItemId = Global.nextListItemId == null ? 1 : Global.nextListItemId + 1;
  return `(${Global.nextListItemId})`;
}

function _nextSnapshotHint() {
  Global.nextSnapshotId = Global.nextSnapshotId == null ? 1 : Global.nextSnapshotId + 1;
  return `[-> check snapshot ${Global.nextSnapshotId}]`;
}
