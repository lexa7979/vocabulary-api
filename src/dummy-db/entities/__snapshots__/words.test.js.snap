// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test-data service "words" (1) has an [1;37masync[0m function [1;37mgetWord()[0m which - when used with valid ID - [0;32mresolves[0m as expected [-> check snapshot 1]: > word1 < 1`] = `
{
  "classId": "(ID:verb)",
  "id": "(ID:word1)",
}
`;

exports[`Test-data service "words" (2) has an [1;37masync[0m function [1;37mgetWordWithTranslations[0m which - when used with valid ID - [0;32mresolves[0m as expected [-> check snapshot 2]: > word2 + translations < 1`] = `
{
  "classId": "(ID:noun)",
  "id": "(ID:word2)",
  "translations": [
    {
      "flectionId": "(ID:noun-flection1)",
      "id": "(ID:word2-translation1)",
      "text_de": "ein Haus",
      "text_sv": "ett hus",
      "wordId": "(ID:word2)",
    },
    {
      "flectionId": "(ID:noun-flection2)",
      "id": "(ID:word2-translation2)",
      "text_de": "das Haus",
      "text_sv": "huset",
      "wordId": "(ID:word2)",
    },
    {
      "flectionId": "(ID:noun-flection3)",
      "id": "(ID:word2-translation3)",
      "text_de": "viele HÃ¤user",
      "text_sv": "mÃ¥nga hus",
      "wordId": "(ID:word2)",
    },
    {
      "flectionId": "(ID:noun-flection4)",
      "id": "(ID:word2-translation4)",
      "text_de": "die HÃ¤user",
      "text_sv": "husen",
      "wordId": "(ID:word2)",
    },
  ],
}
`;

exports[`Test-data service "words" (3) has an [1;37masync[0m function [1;37mlistAllWords()[0m which - when used w/o arguments - [0;32mresolves[0m as expected [-> check snapshot 3]: > all words < 1`] = `
[
  {
    "classId": "(ID:verb)",
    "id": "(ID:word1)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word2)",
  },
  {
    "classId": "(ID:adjective)",
    "id": "(ID:word3)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word4)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word5)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word6)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word7)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word8)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word9)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word10)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word11)",
  },
  {
    "classId": "(ID:noun)",
    "id": "(ID:word12)",
  },
]
`;

exports[`Test-data service "words" (4) has an [1;37masync[0m function [1;37mgetTranslation()[0m which - when used with valid ID - [0;32mresolves[0m as expected [-> check snapshot 4]: > 1st translation of word1 < 1`] = `
{
  "flectionId": "(ID:verb-flection1)",
  "id": "(ID:word1-translation1)",
  "text_de": "gehen",
  "text_sv": "gÃ¥",
  "wordId": "(ID:word1)",
}
`;

exports[`Test-data service "words" (5) has an [1;37masync[0m function [1;37mlistAllTranslationsOfWord()[0m which - when used with valid ID - [0;32mresolves[0m as expected [-> check snapshot 5]: > all translations of word1 < 1`] = `
[
  {
    "flectionId": "(ID:verb-flection1)",
    "id": "(ID:word1-translation1)",
    "text_de": "gehen",
    "text_sv": "gÃ¥",
    "wordId": "(ID:word1)",
  },
  {
    "flectionId": "(ID:verb-flection2)",
    "id": "(ID:word1-translation2)",
    "text_de": "er geht",
    "text_sv": "han gÃ¥r",
    "wordId": "(ID:word1)",
  },
  {
    "flectionId": "(ID:verb-flection2)",
    "id": "(ID:word1-translation3)",
    "text_de": "sie geht",
    "text_sv": "hon gÃ¥r",
    "wordId": "(ID:word1)",
  },
  {
    "flectionId": "(ID:verb-flection2)",
    "id": "(ID:word1-translation4)",
    "text_de": "wir gehen",
    "text_sv": "vi gÃ¥r",
    "wordId": "(ID:word1)",
  },
  {
    "flectionId": "(ID:verb-flection3)",
    "id": "(ID:word1-translation5)",
    "text_de": "er ging",
    "text_sv": "han gick",
    "wordId": "(ID:word1)",
  },
  {
    "flectionId": "(ID:verb-flection4)",
    "id": "(ID:word1-translation6)",
    "text_de": "er ist gegangen",
    "text_sv": "han har gÃ¥tt",
    "wordId": "(ID:word1)",
  },
  {
    "flectionId": "(ID:verb-flection5)",
    "id": "(ID:word1-translation7)",
    "text_de": "geh!|geh",
    "text_sv": "gÃ¥!|gÃ¥",
    "wordId": "(ID:word1)",
  },
]
`;
